easydock/preparation_for_docking.py

```diff
@@ -35,43 +38,241 @@ def mol_from_smi_or_molblock(ligand_string):
         params.useRandomCoords = useRandomCoords
         params.randomSeed = randomSeed
-        conf_stat = AllChem.EmbedMolecule(mole, params)
+        if has_saturated_ring:
+            #10 is used as default according to the C language documentation iirc, but I have to specify the numbers.
+            conf_stat = AllChem.EmbedMultipleConfs(mole, 10, params)
+        else:
+            conf_stat = AllChem.EmbedMolecule(mole, params)
         return mole, conf_stat
-
+    
+    def remove_confs_rms(mol: Chem.Mol, saturated_ring_list: list[list[int]], rms: float=0.25, keep_nconf: Optional[int]=None) -> Chem.Mol:
+        """
+        The function uses AgglomerativeClustering to select conformers.
+
+        :param mol: input molecule with multiple conformers
+        :param rms: discard conformers which are closer than given value to a kept conformer
+        :param keep_nconf: keep at most the given number of conformers. This parameter has precedence over rms
+        :return:
+        """
+
+        def gen_ids(ids: int) -> Iterator[int]:
+            for i in range(1, len(ids)):
+                for j in range(0, i):
+                    yield j, i
+
+        if keep_nconf and mol.GetNumConformers() <= keep_nconf:
+            return mol
+
+        if mol.GetNumConformers() <= 1:
+            return mol
+
+        mol_tmp = Chem.RemoveHs(mol)   # calc rms for heavy atoms only
+        rms_ = GetConformerRMSMatrixForSaturatedRingMolecule(mol_tmp, atomIds=saturated_ring_list, prealigned=False)
+
+        cids = [c.GetId() for c in mol_tmp.GetConformers()]
+        arr = np.zeros((len(cids), len(cids)))
+        for (i, j), v in zip(gen_ids(cids), rms_):
+            arr[i, j] = v
+            arr[j, i] = v
+
+        cl = AgglomerativeClustering(n_clusters=None, linkage='complete', metric='precomputed', distance_threshold=rms).fit(arr)
+
+        keep_ids = []
+        for i in set(cl.labels_):
+            ids = np.where(cl.labels_ == i)[0]
+            j = arr[np.ix_(ids, ids)].mean(axis=0).argmin()
+            keep_ids.append(cids[j])
+        remove_ids = set(cids) - set(keep_ids)
+
+        for cid in sorted(remove_ids, reverse=True):
+            mol.RemoveConformer(cid)
+
+        if mol.GetNumConformers() == 1:
+            return mol
+        
+        if keep_nconf and mol.GetNumConformers() > keep_nconf:
+            arr = arr[np.ix_(keep_ids, keep_ids)]
+            cl = AgglomerativeClustering(n_clusters=keep_nconf, linkage='complete', metric='precomputed').fit(arr)
+
+            keep_ids_nconf_filter = []
+            for i in set(cl.labels_):
+                ids = np.where(cl.labels_ == i)[0]
+                j = arr[np.ix_(ids, ids)].mean(axis=0).argmin()
+                keep_ids_nconf_filter.append(cids[j])
+            remove_ids = set(keep_ids) - set(keep_ids_nconf_filter)
+
+            for cid in sorted(remove_ids, reverse=True):
+                mol.RemoveConformer(cid)
+        
+        return mol
```

Comment:
You indices may not work due to the same issue. And there will be incorrect `cids` on the second step, because it should be recreated. Below I suggest an alternative which might work :) There is a need to chose proper indexing using boolean arrays. I also remove `mol_tmp` object as it is used only once.

```python
rms_ = GetConformerRMSMatrixForSaturatedRingMolecule(Chem.RemoveHs(mol), atomIds=saturated_ring_list, prealigned=False)

cids = [c.GetId() for c in mol.GetConformers()]
arr = np.zeros((len(cids), len(cids)))
for (i, j), v in zip(gen_ids(cids), rms_):
    arr[i, j] = v
    arr[j, i] = v

cl = AgglomerativeClustering(n_clusters=None, linkage='complete', metric='precomputed', distance_threshold=rms).fit(arr)

keep_ids = []
for i in set(cl.labels_):
    ids = np.where(cl.labels_ == i)[0]
    j = arr[np.ix_(ids, ids)].mean(axis=0).argmin()
    keep_ids.append(cids[j])
remove_ids = set(cids) - set(keep_ids)

for cid in sorted(remove_ids, reverse=True):
    mol.RemoveConformer(cid)

if keep_nconf and mol.GetNumConformers() > keep_nconf and mol.GetNumConformers() > 1:
    ids = np.in1d(cids, keep_ids)
    arr = arr[np.ix_(ids, ids)]   # here other indexing operation should be used, because ids is a boolean array
    cl = AgglomerativeClustering(n_clusters=keep_nconf, linkage='complete', metric='precomputed').fit(arr)

    keep_ids = []
    cids = [c.GetId() for c in mol.GetConformers()]
    for i in set(cl.labels_):
        ids = np.where(cl.labels_ == i)[0]
        j = arr[np.ix_(ids, ids)].mean(axis=0).argmin()
        keep_ids.append(cids[j])
    remove_ids = set(cids) - set(keep_ids)

    for cid in sorted(remove_ids, reverse=True):
        mol.RemoveConformer(cid)

return mol
```

Reply:
so you are saying that there is no guarantee that the cids are consecutive? for example, instead of `[0 1 2 3]`, the cids value can be `[1 2 5 6]` and the keep ids can be `[5, 6]`.  In that case, shouldn't boolean array like `ids` sufficient, since it would still take the third and last element of `[1 2 5 6]`

Reply:
Yes, boolean array `ids` should be sufficient to subset an array of RMSD values, like in my example. Therefore, it is not necessary to convert it to explicit indices with `where`. I just did not remember how to subset numpy arrays with logical indices.